### DISCUSSION TITLE
Error with Sparse in Reconstruction Process

### DISCUSSION THREAD

**Discussion by Savillian** [mtex-toolbox/mtex#1240] [Ask Anything] [answered]
Hi everyone,

I've been encountering an issue with the with the job.calcTPVotes command in the parentGrainReconstructor, and wanted to see if anyone knew what was causing the issue? I recall seeing something similar to this issue in the past, but can't find it on the forum.

The job.calcTPVotes command will run to completion and then return the following error. The rest of my code is included above for reference.

```
%% Clean Data
ebsd = ebsd(ebsd.prop.ci>0.1);
ebsd = ebsd('Titanium (Alpha)');
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd('Titanium (Alpha)'),'angle',2*degree);
ebsd(grains(grains.grainSize<2)) = []; % Needs to be small due to the fine scale of the martensite.
[grains,ebsd.grainId] = calcGrains(ebsd('Titanium (Alpha)'),'angle',2*degree,'removeQuadruplePoints');
F = splineFilter;
ebsd = smooth(ebsd('Titanium (Alpha)'),F,'fill');
[grains,ebsd.grainId] = calcGrains(ebsd('Titanium (Alpha)'),'angle',2*degree);

%% Begin Reconstruction
setMTEXpref('xAxisDirection','east'); % east for BF up, west for BF down
setMTEXpref('zAxisDirection','intoPlane');
beta2alpha = orientation.Burgers(ebsd('Titanium (Beta)').CS,ebsd('Titanium (Alpha)').CS);
round2Miller(beta2alpha);
job = parentGrainReconstructor(ebsd, grains);
job.p2c = beta2alpha

% Vote on Triple Points
job.calcTPVotes('minFit',2.5*degree,'maxFit',5*degree)
``` 

_Error using sparse
Sparse matrix sizes must be non-negative integers less than MAXSIZE as defined by COMPUTER.  Use HELP COMPUTER for more
details.

Error in accumVotes (line 53)
  W = full(sparse(grainId(:), parentId(:), prob, maxGrainId, max(parentId(:))));

Error in parentGrainReconstructor/calcTPVotes (line 45)
job.votes = accumVotes(grainInd, parentId, fit, length(job.grains), varargin{:});_

I'd be happy to discuss further if anyone has insight on the matter. Thanks in advance!
Alec

**Comment by nyyssont** [comment]
Happy to discuss, although I don't have a solution.

The error message seems to indicate that there's something wrong with the size specification, the two last variables in the command. One of them, based on the name, seems to be the largest grainId in your ebsd map.

Just guessing: I notice that you have quite a few rounds of calculating grains and cleaning the map. Could it be that the ebsd object retains some grainIds from previous grain calculations, and these give a faulty maxGrainId? I have to admit, I forget if the old grainIds are removed when a new grains object is calculated. In any case, you could try:

```
ebsd.grainId = zeros(length(ebsd),1);
```

prior to your last `calcGrains` step. Does that make any difference?

**Reply by Savillian** [reply]
Thanks for the prompt reply! I suspect a leftover grain calculation is impacting these results also, though I have yet to isolate which one is the problem. 

Applying the zeros matrix did not solve the problem unfortunately. I'll keep troubleshooting to see if it's a grains variable or ebsd variable causing the problem.

**Reply by Savillian** [reply]
I think the cause of the issue lies in the splineFilter. Processing the data via a splineFilter returns the following error: 

```
In EBSD/smooth (line 61) 
Warning: Hexagonal grids are not yet fully supportet for the splineFilter. It might give reasonable
results anyway 
```

If I don't process the data using a splineFilter, the TPCvoting process works without issue.

**Comment by Savillian** [comment, marked as answer]
Hi everyone,

After much trial and tribulation, I found a workaround for this issue. Though I couldn't peg the exact cause, the challenge seemed to originate from the reconstruction voting algorithm being unable to pull out voting criterion from my selected scans. Other EBSD maps of similar materials were able to be reconstructed without issue, and I'm not sure what is different between the two sets.

Regardless, the workaround is to use the job.calcParentFromGraph **[https://mtex-toolbox.github.io/MaParentGrainReconstruction.html]** reconstruction process and not the voting method. This easily reconstructed the parent microstructures and avoided the Sparse error encountered previously.

Hope this helps people in the future!
